### PR TITLE
Support current Azure credential management

### PR DESCRIPTION
**Pull request by malancas**
#### Summary

After debugging authentication failures against ACR from an AKS cluster, I found that the go-containerregistry library uses an Azure credential helper that uses an old version of the Azure identity library and only supports service principals. This adds support for the current Azure credential management system.

This should address https://github.com/sigstore/policy-controller/issues/1391.

### CODE REVIEW

**Review by cpanato — COMMENTED** [review]
lgtm so far

File: pkg/webhook/registryauth/registryauth.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+The K8s auth key chain constructor in go-containerregistry uses an old Azure credential handler.
+We should eventually try to get the Azure credential handler updated upstream in
+go-containerregistry and remove this file. But for now, this custom constructor
+should fix authentication errors encountered when using the policy controller
+with ACR and AKS clusters.
+*/
+var amazonKeychain authn.Keychain = authn.NewKeychainFromHelper(ecr.NewECRHelper(ecr.WithLogger(io.Discard)))
+
+func K8sChainWithCustomACRHelper(ctx context.Context, client kubernetes.Interface, opt k8schain.Options) (authn.Keychain, error) {
```
Comment (hectorj2f) [inline pkg/webhook/registryauth/registryauth.go:47]: Could we rename this function? For a moment I thought we only supported ACS but I can see we also support Amazon and google too.

Comment (malancas) [reply pkg/webhook/registryauth/registryauth.go:47]: I'll rename it to `NewK8sKeyChain`. The comment at the top of the file should give the reader further explanation why we aren't using the one in go-containerregsitry.

**Review by hectorj2f — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -71,6 +71,68 @@ If you would like to use the local Kind registry instead of a live one,
...
+See the [official documentation](https://learn.microsoft.com/en-us/azure/container-registry/authenticate-kubernetes-options#scenarios).
+
+### Managed Identities for AKS Clusters
+
+See the [official documentation](https://learn.microsoft.com/en-us/azure/aks/cluster-container-registry-integration?toc=%2Fazure%2Fcontainer-registry%2Ftoc.json&bc=%2Fazure%2Fcontainer-registry%2Fbreadcrumb%2Ftoc.json&tabs=azure-cli) for more details.
+
+1. You must enable managed identities for the cluster using the `--enable-managed-identities` flag with either the `az aks create` or `az aks update` commands
+1. You must attach the ACR to the AKS cluster using the `--attach-acr` with either
+the `az aks create` or `az aks update` commands. See [here](https://learn.microsoft.com/en-us/azure/aks/cluster-container-registry-integration?toc=%2Fazure%2Fcontainer-registry%2Ftoc.json&bc=%2Fazure%2Fcontainer-registry%2Fbreadcrumb%2Ftoc.json&tabs=azure-cli#create-a-new-aks-cluster-and-integrate-with-an-existing-acr) for more details
+1. You must set the `AZ_CLIENT_ID` environment variable to the managed identity's client ID.
```
Comment (ejohn20) [inline README.md:92]: Should `AZ_CLIENT_ID` be `AZURE_CLIENT_ID`?

File: README.md [hunk 3]
```diff
@@ -71,6 +71,68 @@ If you would like to use the local Kind registry instead of a live one,
...
+This will detected by the Azure credential manager
+
+When you create a cluster that has managed identities enabled,
+a user assigned managed identity called
+`<AKS cluster name>-agentpool`. Use this identity's client ID
+when setting `AZ_CLIENT_ID`. Make sure the ACR is attached to
```
Comment (ejohn20) [inline README.md:98]: Should `AZ_CLIENT_ID` be `AZURE_CLIENT_ID`?

File: README.md [hunk 4]
```diff
@@ -71,6 +71,68 @@ If you would like to use the local Kind registry instead of a live one,
...
+your cluster.
+
+#### Installing Policy Controller from this repository
+
+If you are deploying policy-controller directly from this repository with
+`make ko-apply`, you will need to add `AZ_CLIENT_ID` to the list of environment
```
Comment (ejohn20) [inline README.md:104]: Should `AZ_CLIENT_ID` be `AZURE_CLIENT_ID`?

File: README.md [hunk 5]
```diff
@@ -71,6 +71,68 @@ If you would like to use the local Kind registry instead of a live one,
...
+variables in the [webhook deployment configuration](config/webhook.yaml).
+
+#### Installing Policy Controller from the Helm chart
+
+You can provide the managed identity's client ID as a custom environment
+variable when installing the Helm chart:
+
+```bash
+helm upgrade --install policy-controller sigstore/policy-controller --version 0.9.0 \
+--set webhook.env.AZ_CLIENT_ID=my-managed-id-client-id
```
Comment (ejohn20) [inline README.md:114]: Should `AZ_CLIENT_ID` be `AZURE_CLIENT_ID`?

File: README.md [hunk 6]
```diff
@@ -71,6 +71,68 @@ If you would like to use the local Kind registry instead of a live one,
...
+```
+
+### Service Principals for AKS Clusters
+
+#### Installing Policy Controller from this repository
+
+If you are deploying policy-controller directly from this repository with
+`make ko-apply`, you will need to add `AZ_CLIENT_ID` and `AZURE_TENANT_ID` to
```
Comment (ejohn20) [inline README.md:122]: Should `AZ_CLIENT_ID` be `AZURE_CLIENT_ID`?

**Review by ejohn20 — COMMENTED** [review]
Looking through these docs, shouldn't the `AZ_CLIENT_ID` env var be `AZURE_CLIENT_ID`?

**Comment by malancas** [comment]
> Looking through these docs, shouldn't the `AZ_CLIENT_ID` env var be `AZURE_CLIENT_ID`?

Yes, meant to update those earlier. They are updated now.

**Review by codysoyland — APPROVED** [review]
Nice job @malancas -- I was able to exercise this branch against an AKS cluster using managed identities.

**Review by hectorj2f — APPROVED** [review]
LGTM

**Pull request merged**